bentoml/_internal/io_descriptors/base.py

```diff
@@ -1,44 +1,86 @@
 import typing as t
 from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING
 
 from starlette.requests import Request
 from starlette.responses import Response
 
+from ..types import FileLike
+
+if TYPE_CHECKING:  # pragma: no cover
+    import numpy as np  # noqa: F401
+    import pandas as pd  # noqa: F401
+    import PIL.Image  # noqa: F401
+
+
+JSONType = t.Union[str, t.Dict[str, t.Any]]
+
+ImageType = t.Union["PIL.Image.Image", "np.ndarray[t.Any, np.dtype[t.Any]]"]
+
+IOType = t.Union[
+    str,
+    JSONType,
+    FileLike,
+    ImageType,
+    "np.ndarray[t.Any, np.dtype[t.Any]]",
+    "pd.DataFrame",
+    "pd.Series[t.Any]",
+]
+
+
 IOPyObj = t.TypeVar("IOPyObj")
 
 
+_T = t.TypeVar("_T")
+
+
 class IODescriptor(ABC, t.Generic[IOPyObj]):
     """
     IODescriptor describes the input/output data format of an InferenceAPI defined
      in a bentoml.Service
     """
 
     HTTP_METHODS = ["POST"]
+    _init_str: str = ""
+
+    def __new__(cls: t.Type[_T], *args: t.Any, **kwargs: t.Any) -> _T:
+        self = super().__new__(cls)
+        arg_strs = tuple(repr(i) for i in args) + tuple(
+            f'{k}={repr(v)}' for k, v in kwargs.items()
+        )
+        setattr(self, "_init_str", f"{cls.__name__}({', '.join(arg_strs)})")
```

Comment:
should we do anything for IO descriptors that are created via `from_sample`?

Reply:
maybe add a attribute `_from_sample` = True?

Reply:
we can save it as an example, the present it on openapi as a $ref to that example

Reply:
that would be really nice